src/core/qgsmaplayer.cpp

```diff
@@ -755,8 +755,15 @@ bool QgsMapLayer::writeLayerXml( QDomElement &layerElement, QDomDocument &docume
 
   if ( !mExtent3D.isNull() && dataProvider() && dataProvider()->elevationProperties() && dataProvider()->elevationProperties()->containsElevationData() )
     layerElement.appendChild( QgsXmlUtils::writeBox3D( mExtent3D, document ) );
-  else if ( !mExtent2D.isNull() )
-    layerElement.appendChild( QgsXmlUtils::writeRectangle( mExtent2D, document ) );
+  else
+  {
+    // Extent might be null because lazily set
+    const QgsRectangle extent2D { mExtent2D.isNull() ? extent() : mExtent2D };
+    if ( !extent2D.isNull() )
+    {
+      layerElement.appendChild( QgsXmlUtils::writeRectangle( mExtent2D, document ) );
```

Comment:
shouldn' that be ?
```suggestion
      layerElement.appendChild( QgsXmlUtils::writeRectangle( extent2D, document ) );
```

Reply:
Actually the outcome is the same because when calling `extent()` the lazy evaluation of the extent is triggered and the mExtent2D is updated with the actual extent. But it's definitely clearer this way.